File: Leetcode/C++/5. Longest Palindromic Substring/solution.cpp
// https://leetcode.com/problems/longest-palindromic-substring/description/

#include <string>

using namespace std;

class Solution
{
public:
    string longestPalindrome(string s)
    {
        int left = 0, right = 0;
        int start_point = 0, max_length = 0;
        for (int idx = 0; idx < s.size(); idx++)
        {
            right = idx;
            // s ~right - 1 = same s[idx] character palindrome
            while (right < s.size() & s[idx] == s[right])
            {
                right += 1;
            }
            left = idx - 1;
            // Find palindrome not same character s[idx]
            while (left >= 0 && right < s.size() && s[left] == s[right])
            {
                left -= 1;
                right += 1;
            }
            // [left + 1 ~ right - 1] is palindrome
            int length = right - left - 1;
            if (length > max_length)
            {
                start_point = left + 1;
                max_length = length;
            }
        }
        return s.substr(start_point, max_length);
    }
};

File: Leetcode/Java/5. Longest Palindromic Substring/solution.java
// https://leetcode.com/problems/longest-palindromic-substring/description/

class Solution {
    public String longestPalindrome(String s) {
        int maxSize = 0;
        int palinIndex = 0;

        for (int idx = 0; idx < s.length(); idx++){
            int left = idx - 1;
            int right = idx;

            // Find same s[idx] character palindrome
            while (right < s.length() && s.charAt(right) == s.charAt(idx)) {
                right += 1;
            }

            // Find palindrome not same character s[idx]
            while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)){
                left -= 1;
                right += 1;
            }
            // [left + 1 ~ right - 1] is palindrome
            if (right - left - 1 > maxSize){
                maxSize = right - left - 1;
                palinIndex = left + 1;
            }
        }

        return s.substring(palinIndex, palinIndex + maxSize);
    }
}